# Proyecto Sprint 3 - Data Wrangling
En este proyecto se trabajó con la prueba de hipótesis para analizar la actividad de escucha de música de usuarios en dos diferentes ciudades, y analizar la diferencia de actividad entre ellas. 
El proyecto se divide en 3 etapas: en la primera se describen los datos de nuestros datasets a primera vista, en la segunda se procesan los datos para poderlos analizar, y en la tercera se lleva a cabo la prueba de hipótesis.
También se exploró el uso de un índice de contenidos enlazado a cada sección.

# Déjame escuchar la música

# Contenido <a id='back'></a>

* [Introducción](#intro)
* [Etapa 1. Descripción de los datos](#data_review)
    * [Conclusiones](#data_review_conclusions)
* [Etapa 2. Preprocesamiento de datos](#data_preprocessing)
    * [2.1 Estilo del encabezado](#header_style)
    * [2.2 Valores ausentes](#missing_values)
    * [2.3 Duplicados](#duplicates)
    * [2.4 Conclusiones](#data_preprocessing_conclusions)
* [Etapa 3. Prueba de hipótesis](#hypothesis)
    * [3.1 Hipótesis 1: actividad de los usuarios y las usuarias en las dos ciudades](#activity)
* [Conclusiones](#end)

# Introducción <a id='intro'></a> 

En este proyecto, compararás las preferencias musicales de las ciudades de Springfield y Shelbyville. Estudiarás datos reales de transmisión de música online para probar la hipótesis a continuación y comparar el comportamiento de los usuarios y las usuarias de estas dos ciudades.






### Objetivo:
Prueba la hipótesis:
1. La actividad de los usuarios y las usuarias difiere según el día de la semana y dependiendo de la ciudad.

### Etapas
Los datos del comportamiento del usuario se almacenan en el archivo `/Users/RPA/Downloads/music_project_en.csv`. No hay ninguna información sobre la calidad de los datos, así que necesitarás examinarlos antes de probar la hipótesis.

Primero, evaluarás la calidad de los datos y verás si los problemas son significativos. Entonces, durante el preprocesamiento de datos, tomarás en cuenta los problemas más críticos.

Tu proyecto consistirá en tres etapas:
 1. Descripción de los datos.
 2. Preprocesamiento de datos.
 3. Prueba de hipótesis.
 4. Conclusiones

[Volver a Contenidos](#back)

# Etapa 1. Descripción de los datos <a id='data_review'></a>

In [140]:
# Importar librerías y cargar archivos
import pandas as pd
df = pd.read_csv('/Users/RPA/Downloads/music_project_en.csv')

In [141]:
# Obtener las 10 primeras filas de la tabla df
print(df.head(10))
# Obtener la información general sobre nuestros datos
print(df.info())

     userID                        Track            artist   genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile    rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg    rock   
2    20EC38            Funiculì funiculà       Mario Lanza     pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice    folk   
4  E2DC1FAE                  Soul People        Space Echo   dance   
5  842029A1                       Chains          Obladaet  rusrap   
6  4CB90AA5                         True      Roman Messer   dance   
7  F03E1C1F             Feeling This Way   Polina Griffith   dance   
8  8FA1D3BE                     L’estate       Julia Dalia  ruspop   
9  E772D5C0                    Pessimist               NaN   dance   

        City        time        Day  
0  Shelbyville  20:28:33  Wednesday  
1  Springfield  14:07:09     Friday  
2  Shelbyville  20:58:07  Wednesday  
3  Shelbyville  08:37:09     Monday  
4  Springfield  08:34:34     Monday  
5

Estas son nuestras observaciones sobre la tabla:
Contiene siete columnas que almacenan los mismos tipos de datos: `object`.

Según la documentación:
- `' userID'`: identificador del usuario o la usuaria;
- `'Track'`: título de la canción;
- `'artist'`: nombre del artista;
- `'genre'`: género de la pista;
- `'City'`: ciudad del usuario o la usuaria;
- `'time'`: la hora exacta en la que se reprodujo la canción;
- `'Day'`: día de la semana.

Podemos ver tres problemas con el estilo en los encabezados de la tabla:
1. Algunos encabezados están en mayúsculas, otros en minúsculas.
2. Hay espacios en algunos encabezados.
3. Tercer problema: algunos nombres de las columnas no son muy descriptivos, como "Day", no se sabe a simple vista si es dia de la semana o del mes, o "time" podria ser la hora en la que se escucho la cancion, o el tiempo de duracion de la cancion. Otro problema podria ser que userID no esta en forma snake case.



## Observaciones: <a id='data_review_conclusions'></a>

Podemos observar que los tipos de datos en las filas son "object", que en pandas podría ser cualquier dato que no sea int, float o bool. Para saber qué almacena cada columna, podemos ver la documentacion y quizas editar los nombres de las columnas para que sea más claro a simple vista qué contienen. 
Mientras que sé hay informacion útil en los datos que tenemos por ahora, es necesario indagar un poco más para probar nuestra hipótesis. 
Viendo la información de nuestro DF, podemos ver que hay tres columnas con datos vacíos, lo cual podría ser clave a la hora de querer comprobar nuestra hipótesis. Hasta ahora, no se han visto filas duplicadas, pero eso no significa que no existan dentro de todo el DF. 

[Volver a Contenidos](#back)

# Etapa 2. Preprocesamiento de datos <a id='data_preprocessing'></a>

El objetivo aquí es preparar los datos para que sean analizados.
El primer paso es resolver cualquier problema con los encabezados. Luego podemos avanzar a los valores ausentes y duplicados.


## Estilo del encabezado <a id='header_style'></a>
Muestra los encabezados de la tabla (los nombres de las columnas):

In [142]:
# Muestra los nombres de las columnas
print(df.columns)

Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='object')


Cambia los encabezados de la tabla de acuerdo con las reglas del buen estilo:
* Todos los caracteres deben ser minúsculas.
* Elimina los espacios.
* Si el nombre tiene varias palabras, utiliza snake_case.

In [143]:
# Bucle en los encabezados poniendo todo en minúsculas
lower_case_columns = []
for name in df.columns:
    lowered_name = name.lower()
    lower_case_columns.append(lowered_name)
df.columns = lower_case_columns

# Bucle en los encabezados eliminando los espacios
stripped_names = []
for name in df.columns:
    strip_name = name.strip()
    stripped_names.append(strip_name)
df.columns = stripped_names

# Cambiar el nombre de la columna "userid"
snake_case = {'userid':'user_id'}
df = df.rename(columns = snake_case,)
print(df.columns)

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


[Volver a Contenidos](#back)

## Valores ausentes <a id='missing_values'></a>

In [144]:
# Calcular el número de valores ausentes
print(df.isna().sum())

user_id       0
track      1343
artist     7567
genre      1198
city          0
time          0
day           0
dtype: int64


In [145]:
# Bucle en los encabezados reemplazando los valores ausentes con 'unknown'
columns_to_replace = ['track', 'artist', 'genre']
for col in columns_to_replace:
    df[col] = df[col].fillna('unknown')

In [146]:
# Contar valores ausentes
print(df.isna().sum())

user_id    0
track      0
artist     0
genre      0
city       0
time       0
day        0
dtype: int64


[Volver a Contenidos](#back)

## Duplicados <a id='duplicates'></a>
Encuentra el número de duplicados explícitos en la tabla. Una vez más, debes aplicar dos métodos en una secuencia para obtener la cantidad de duplicados explícitos.

In [147]:
# Contar duplicados explícitos
print(df.duplicated().sum())

3826


In [148]:
# Eliminar duplicados explícitos
df = df.drop_duplicates()
print(df.duplicated().sum())

0


Ahora queremos deshacernos de los duplicados implícitos en la columna `genre`. Por ejemplo, el nombre de un género se puede escribir de varias formas. Dichos errores también pueden afectar al resultado.

Para hacerlo, primero mostremos una lista de nombres de género únicos, ordenados en orden alfabético. Para ello:
* Extrae la columna `genre` del DataFrame.
* Llama al método que devolverá todos los valores únicos en la columna extraída.


In [149]:
# Inspeccionar los nombres de géneros únicos
print(df['genre'].unique())

['rock' 'pop' 'folk' 'dance' 'rusrap' 'ruspop' 'world' 'electronic'
 'unknown' 'alternative' 'children' 'rnb' 'hip' 'jazz' 'postrock' 'latin'
 'classical' 'metal' 'reggae' 'triphop' 'blues' 'instrumental' 'rusrock'
 'dnb' 'türk' 'post' 'country' 'psychedelic' 'conjazz' 'indie'
 'posthardcore' 'local' 'avantgarde' 'punk' 'videogame' 'techno' 'house'
 'christmas' 'melodic' 'caucasian' 'reggaeton' 'soundtrack' 'singer' 'ska'
 'salsa' 'ambient' 'film' 'western' 'rap' 'beats' "hard'n'heavy"
 'progmetal' 'minimal' 'tropical' 'contemporary' 'new' 'soul' 'holiday'
 'german' 'jpop' 'spiritual' 'urban' 'gospel' 'nujazz' 'folkmetal'
 'trance' 'miscellaneous' 'anime' 'hardcore' 'progressive' 'korean'
 'numetal' 'vocal' 'estrada' 'tango' 'loungeelectronic' 'classicmetal'
 'dubstep' 'club' 'deep' 'southern' 'black' 'folkrock' 'fitness' 'french'
 'disco' 'religious' 'hiphop' 'drum' 'extrememetal' 'türkçe'
 'experimental' 'easy' 'metalcore' 'modern' 'argentinetango' 'old' 'swing'
 'breaks' 'eurofolk' 

Busca en la lista para encontrar duplicados implícitos del género `hiphop`. Estos pueden ser nombres escritos incorrectamente o nombres alternativos para el mismo género.

Verás los siguientes duplicados implícitos:
* `hip`
* `hop`
* `hip-hop`

Para deshacerte de ellos, crea una función llamada `replace_wrong_genres()` con dos parámetros:
* `wrong_genres=`: esta es una lista que contiene todos los valores que necesitas reemplazar.
* `correct_genre=`: este es un string que vas a utilizar como reemplazo.

Como resultado, la función debería corregir los nombres en la columna `'genre'` de la tabla `df`, es decir, remplazar cada valor de la lista `wrong_genres` por el valor en `correct_genre`.

Dentro del cuerpo de la función, utiliza un bucle `'for'` para iterar sobre la lista de géneros incorrectos, extrae la columna `'genre'` y aplica el método `replace` para hacer correcciones.

In [150]:
# Función para reemplazar duplicados implícitos
def replace_wrong_genres(df, column, wrong_genres, correct_genre):
    for wrong_genre in wrong_genres:
        df[column] = df[column].replace(wrong_genres, correct_genre)
    return df
# Eliminar duplicados implícitos
wrong_genres = ['hip', 'hop', 'hip-hop']
correct_genre = 'hiphop'
df = replace_wrong_genres(df, 'genre', wrong_genres, correct_genre)
print(df)

        user_id                              track            artist  \
0      FFB692EC                  Kamigata To Boots  The Mass Missile   
1      55204538        Delayed Because of Accident  Andreas Rönnberg   
2        20EC38                  Funiculì funiculà       Mario Lanza   
3      A3DD03C9              Dragons in the Sunset        Fire + Ice   
4      E2DC1FAE                        Soul People        Space Echo   
...         ...                                ...               ...   
65074  729CBB09                            My Name            McLean   
65075  D08D4A55  Maybe One Day (feat. Black Spade)       Blu & Exile   
65076  C5E3A0D5                          Jalopiina           unknown   
65077  321D0506                      Freight Train     Chas McDevitt   
65078  3A64EF84          Tell Me Sweet Little Lies      Monica Lopez   

            genre         city      time        day  
0            rock  Shelbyville  20:28:33  Wednesday  
1            rock  Springfi

Asegúrate de que los nombres duplicados han sido eliminados. Muestra la lista de valores únicos de la columna `'genre'` una vez más:

In [151]:
# Comprobación de duplicados implícitos
print(df['genre'].unique())

['rock' 'pop' 'folk' 'dance' 'rusrap' 'ruspop' 'world' 'electronic'
 'unknown' 'alternative' 'children' 'rnb' 'hiphop' 'jazz' 'postrock'
 'latin' 'classical' 'metal' 'reggae' 'triphop' 'blues' 'instrumental'
 'rusrock' 'dnb' 'türk' 'post' 'country' 'psychedelic' 'conjazz' 'indie'
 'posthardcore' 'local' 'avantgarde' 'punk' 'videogame' 'techno' 'house'
 'christmas' 'melodic' 'caucasian' 'reggaeton' 'soundtrack' 'singer' 'ska'
 'salsa' 'ambient' 'film' 'western' 'rap' 'beats' "hard'n'heavy"
 'progmetal' 'minimal' 'tropical' 'contemporary' 'new' 'soul' 'holiday'
 'german' 'jpop' 'spiritual' 'urban' 'gospel' 'nujazz' 'folkmetal'
 'trance' 'miscellaneous' 'anime' 'hardcore' 'progressive' 'korean'
 'numetal' 'vocal' 'estrada' 'tango' 'loungeelectronic' 'classicmetal'
 'dubstep' 'club' 'deep' 'southern' 'black' 'folkrock' 'fitness' 'french'
 'disco' 'religious' 'drum' 'extrememetal' 'türkçe' 'experimental' 'easy'
 'metalcore' 'modern' 'argentinetango' 'old' 'swing' 'breaks' 'eurofolk'
 'stone

[Volver a Contenidos](#back)

## Tus observaciones <a id='data_preprocessing_conclusions'></a>

Se encontraron 3,000+ datos explícitos duplicados, los cuales fueron encontrados usando .duplicated() y eliminados usando drop_duplicates. Una vez hecho esto, se analizaron los duplicados implícitos (usando .unique()), donde se encontró un género musical que estaba escrito de 4 maneras diferentes. Se utilizó el método replace dentro de un for loop en una funcion para convertir esos 4 nombres en uno solo. Una vez hecho eso, se comprobó una última vez que no hubiera duplicados. Los resultados fueron un DF sin filas duplicadas, y con la columna 'genre' teniendo un solo valor único para el genero hiphop.  

[Volver a Contenidos](#back)

# Etapa 3. Prueba de hipótesis <a id='hypothesis'></a>

## Hipótesis: comparar el comportamiento del usuario o la usuaria en las dos ciudades <a id='activity'></a>

La hipótesis afirma que existen diferencias en la forma en que los usuarios y las usuarias de Springfield y Shelbyville consumen música. Para comprobar esto, usa los datos de tres días de la semana: lunes, miércoles y viernes.

* Agrupa a los usuarios y las usuarias por ciudad.
* Compara el número de canciones que cada grupo reprodujo el lunes, el miércoles y el viernes.


Realiza cada cálculo por separado.

El primer paso es evaluar la actividad del usuario en cada ciudad. Recuerda las etapas dividir-aplicar-combinar de las que hablamos anteriormente en la lección. Tu objetivo ahora es agrupar los datos por ciudad, aplicar el método apropiado para contar durante la etapa de aplicación y luego encontrar la cantidad de canciones reproducidas en cada grupo especificando la columna para obtener el recuento.

A continuación se muestra un ejemplo de cómo debería verse el resultado final:
`df.groupby(by='....')['column'].method()`Realiza cada cálculo por separado.

Para evaluar la actividad de los usuarios y las usuarias en cada ciudad, agrupa los datos por ciudad y encuentra la cantidad de canciones reproducidas en cada grupo.



In [152]:
# Contar las canciones reproducidas en cada ciudad
city_df = df.groupby('city')['track'].count()
print(city_df)

city
Shelbyville    18512
Springfield    42741
Name: track, dtype: int64


`Comenta tus observaciones aquí`
Podemos ver que en Shelbyville se escucharon 18,512 canciones; y en Springfield 42,741. Es decir, en Springfield se escucharon mas del doble de canciones que en Shelbyville. 

Ahora agrupemos los datos por día de la semana y encontremos el número de canciones reproducidas el lunes, miércoles y viernes. 


In [153]:
springfield = df[df['city']=='Springfield']
print(springfield.groupby('day')['track'].count().reset_index())

         day  track
0     Friday  15945
1     Monday  15740
2  Wednesday  11056


In [154]:
shelbyville = df[df['city']=='Shelbyville']
print(shelbyville.groupby('day')['track'].count().reset_index())

         day  track
0     Friday   5895
1     Monday   5614
2  Wednesday   7003


Ya sabes cómo contar entradas agrupándolas por ciudad o día. Ahora necesitas escribir una función que pueda contar entradas según ambos criterios simultáneamente.

Crea la función `number_tracks()` para calcular el número de canciones reproducidas en un determinado día **y** ciudad. La función debe aceptar tres parámetros:
- `tu dataframe`
- `day`
- `city`

Dentro de la función, aplicarás un filtrado consecutivo con indexación lógica.

Primero filtra los datos por día y luego filtra la tabla resultante por ciudad.

Después de filtrar los datos por dos criterios, cuenta el número de valores de la columna 'user_id' en la tabla resultante. Este recuento representa el número de entradas que estás buscando. Guarda el resultado en una nueva variable y devuélvelo desde la función.

In [155]:
# Declara la función number_tracks() con dos parámetros: day= y city=.

def number_tracks(df, day, city):
    filtered_by_day = df[df['day'] == day]
    filtered_by_city = filtered_by_day[filtered_by_day['city'] == city]
    filtered_df = filtered_by_city['track'].count()
    return filtered_df

Llama a `number_tracks()` seis veces, cambiando los valores de los parámetros para que recuperes los datos de ambas ciudades para cada uno de los tres días.

In [156]:
# El número de canciones reproducidas en Springfield el lunes
def number_tracks(df, day, city):
    filtered_by_day = df[df['day'] == day]
    filtered_by_city = filtered_by_day[filtered_by_day['city'] == city]
    filtered_df = filtered_by_city['track'].count()
    return filtered_df

day= 'Monday'
city = 'Springfield'
M_Sf = number_tracks(df, day, city)
print(M_Sf)

15740


In [157]:
# El número de canciones reproducidas en Shelbyville el lunes
day = 'Monday'
city = 'Shelbyville'
M_Sv = number_tracks(df, day, city)
print(M_Sv)

5614


In [158]:
# El número de canciones reproducidas en Springfield el miércoles
day = 'Wednesday'
city = 'Springfield'
W_Sf = number_tracks(df, day, city)
print(W_Sf)

11056


In [159]:
# El número de canciones reproducidas en Shelbyville el miércoles
day = 'Wednesday'
city = 'Shelbyville'
W_Sv = number_tracks(df, day, city)
print(W_Sv)

7003


In [160]:
# El número de canciones reproducidas en Springfield el viernes
day = 'Friday'
city = 'Springfield'
F_Sf = number_tracks(df, day, city)
print(F_Sf)

15945


In [161]:
# El número de canciones reproducidas en Shelbyville el viernes
day = 'Friday'
city = 'Shelbyville'
F_Sv = number_tracks(df, day, city)
print(F_Sv)

5895


# **Conclusiones**

'La hipotesis es correcta, ya que podemos ver que el numero de canciones escuchadas difiere entre los dias de la semana, asi como un lunes no se escucha la misma cantidad de canciones en Shelbyville que el mismo lunes en Springfield. Por lo tanto, el decir que "La actividad de los usuarios y las usuarias difiere según el día de la semana y dependiendo de la ciudad" es correcto.' 

[Volver a Contenidos](#back)